### PR TITLE
Marshal placement snapshot as maps instead of arrays

### PR DESCRIPTION
**Pull request by xichen2020**
cc @cw9 @robskillington

### CODE REVIEW

File: placement/placement.go [hunk 1]
```diff
@@ -156,28 +155,30 @@ func (su sortableUInt32) Swap(i, j int) {
...
 }
 
-func convertJSONtoSnapshot(hsjs hostShardsJSONs) (snapshot, error) {
+func convertJSONtoSnapshot(m map[string]hostShardsJSON) (snapshot, error) {
 	var err error
-	hss := make([]HostShards, len(hsjs))
+	hss := make([]HostShards, len(m))
 	shardsReplicaMap := make(map[uint32]int)
-	for i, hsj := range hsjs {
+	i := 0
```
Comment (cw9) [inline placement/placement.go]: is this i used?

File: placement/placement_test.go [hunk 2]
```diff
@@ -160,23 +160,23 @@ func TestValidate(t *testing.T) {
...
-	]`
+	invalidJSON := `{
+		:{"ID":123,"Rack":"r1","Shards":[0,7,11]}
+	}`
 	data := []byte(invalidJSON)
 	ps, err := NewPlacementFromJSON(data)
 	assert.Nil(t, ps)
 	assert.Error(t, err)
 
-	validJSON := `[
```
Comment (cw9) [inline placement/placement_test.go]: the order is no longer guaranteed is it?

File: placement/placement.go (hunk 1 above)
Comment (xichen2020) [inline placement/placement.go]: yeah in the for loop

File: placement/placement_test.go (hunk 2 above)
Comment (xichen2020) [inline placement/placement_test.go]: no it's not, but apparently the diff is smaller with maps if we change the placements

**Comment by cw9** [comment]
LGTM

**Pull request merged**